dev/test/transaction.ts

```diff
@@ -377,75 +423,72 @@ function runTransaction<T>(
       setTimeoutHandler(setTimeout);
       expect(expectedRequests.length).to.equal(
         0,
-        'Missing requests: ' + expectedRequests.map(r => r.type).join(', ')
+        'Missing requests: ' + expectedRequests.map(r => r!.type).join(', ')
       );
     }
   });
 }
 
 describe('successful transactions', () => {
   it('empty transaction', () => {
-    return runTransaction(
-      /* transactionOptions= */ {},
-      () => {
-        return Promise.resolve();
-      },
-      begin(),
-      commit()
-    );
+    return runTransaction(/* transactionOptions= */ {}, () => {
+      return Promise.resolve();
+    });
   });
 
   it('returns value', () => {
-    return runTransaction(
-      /* transactionOptions= */ {},
-      () => {
-        return Promise.resolve('bar');
-      },
-      begin(),
-      commit()
-    ).then(val => {
+    return runTransaction(/* transactionOptions= */ {}, () => {
+      return Promise.resolve('bar');
+    }).then(val => {
       expect(val).to.equal('bar');
     });
   });
 });
 
 describe('failed transactions', () => {
-  const retryBehavior: {[code: number]: boolean} = {
-    [Status.CANCELLED]: true,
-    [Status.UNKNOWN]: true,
-    [Status.INVALID_ARGUMENT]: false,
-    [Status.DEADLINE_EXCEEDED]: true,
-    [Status.NOT_FOUND]: false,
-    [Status.ALREADY_EXISTS]: false,
-    [Status.RESOURCE_EXHAUSTED]: true,
-    [Status.FAILED_PRECONDITION]: false,
-    [Status.ABORTED]: true,
-    [Status.OUT_OF_RANGE]: false,
-    [Status.UNIMPLEMENTED]: false,
-    [Status.INTERNAL]: true,
-    [Status.UNAVAILABLE]: true,
-    [Status.DATA_LOSS]: false,
-    [Status.UNAUTHENTICATED]: true,
+  const retryBehavior: {[code: number]: {retry: boolean; rollback: boolean}} = {
+    [Status.CANCELLED]: {retry: true, rollback: true},
+    [Status.UNKNOWN]: {retry: true, rollback: true},
+    [Status.DEADLINE_EXCEEDED]: {retry: true, rollback: true},
+    [Status.RESOURCE_EXHAUSTED]: {retry: true, rollback: true},
+    [Status.ABORTED]: {retry: true, rollback: false},
+    [Status.INTERNAL]: {retry: true, rollback: true},
+    [Status.UNAVAILABLE]: {retry: true, rollback: true},
+    [Status.UNAUTHENTICATED]: {retry: true, rollback: true},
+
+    [Status.INVALID_ARGUMENT]: {retry: false, rollback: false},
+    [Status.NOT_FOUND]: {retry: false, rollback: true},
+    [Status.ALREADY_EXISTS]: {retry: false, rollback: false},
+    [Status.FAILED_PRECONDITION]: {retry: false, rollback: false},
+    [Status.OUT_OF_RANGE]: {retry: false, rollback: true},
+    [Status.UNIMPLEMENTED]: {retry: false, rollback: true},
+    [Status.DATA_LOSS]: {retry: false, rollback: true},
```

Comment:
Yes I think some thorough discussion is warranted on this. I can reverse it so that it can be implemented separately.

So this behaviour is not documented anywhere, but I discovered it via experimentation with the REST API, I'll show an example below.

But it kind of makes sense, here's an obvious example: if the commit fails saying "transaction has expired", then obviously there is no need to attempt rolling back an expired transaction.

However I've found is that the behaviour extends to several other error codes too.

Here's an example you can try yourself:

(1) Hold a lock on a document (non-existent document)

```
POST /v1/projects/test-project/databases/(default)/documents:batchGet HTTP/1.1
Host: firestore.googleapis.com
Authorization: Bearer •••••
Content-Type: application/json

{
    "newTransaction": {
        "readWrite": {}
    },
    "documents": [
        "projects/test-project/databases/(default)/documents/collection_id/non_existent_doc"
    ]
}
```

(2) Attempt to write that document outside the transaction. You will find that this request is blocked until the transaction from (1) completes for fails or times out.

```
POST /v1/projects/test-project/databases/(default)/documents:batchWrite HTTP/1.1
Host: firestore.googleapis.com
Authorization: Bearer •••••
Content-Type: application/json

{
    "writes": [
        {
            "update": {
                "name": "projects/test-project/databases/(default)/documents/collection_id/non_existent_doc",
                "fields": {}
            }
        }
    ]
}
```

(3) Commit the transaction from (1) with a precondition that will fail (i.e. precondition "exists" when the document doesn't exists causes the commit to fail).

```
POST /v1/projects/test-project/databases/(default)/documents:commit HTTP/1.1
Host: firestore.googleapis.com
Authorization: Bearer •••••
Content-Type: application/json

{
    "transaction": "<the transaction ID from (1)>"
    "writes": [
        {
            "update": {
                "name": "projects/test-project/databases/(default)/documents/collection_id/non_existent_doc",
                "fields": {}
            },
            "currentDocument": {
                "exists": true
            }
        }
    ]
}
```

The commit will fail with "NOT_FOUND" (no document to update). And in that moment, you will see that the request from (2) completes successfully meaning the failed commit released the lock on that document.

You'll need to complete this sequence within 20s or so otherwise (2) may time out anyway.

We'd definitely need input from the appropriate team on which error codes mean that the transaction has been aborted, and whether or not the behaviour is reliable. That is, it needs to be documented!

Reply:
I received from clarification from Firestore. The rollback can be called in a non-blocking fashion. We do not need to wait for result of rollback before continuing.

I think making the rollback non-blocking will achieve the same latency as omitting the callback.

Feel free to implement this.

Reply:
Ok I get that. I'll remove those changes.

> The rollback can be called in a non-blocking fashion. We do not need to wait for result of rollback before continuing.

Hmm, but without knowing the specifics of the RPC or REST transport, what's to say that which of the rollback or the next transaction attempt requests arrive at the backend first? I.e. they could be on different streams or connections? Isn't that asking for a race condition?

Or is it that the next attempt would be blocked at the backend until the rollback is received anyway, so it doesn't matter which arrives first?

What will the backend do if/when it receives a `readWrite: { retryTransaction: "..." }` options to retry a transaction that hasn't been committed or rolled back yet? Does it automatically roll back the provided transaction?

Finally, regarding your point about errors from reads. If the initial read operation rejects, then as it is implemented currently we will obviously have no transaction ID to roll back. I was pondering whether or not it may be possible for:
1. The backend to lazily start the transaction and establish the lock, only for something in the SDK (e.g. validation) to reject, preventing us from receiving the transaction ID, and preventing us from rolling back the transaction that was established. I figured this was both highly unlikely and inescapable. 
2. The backend to send the "transaction ID response" in the response stream and then fail/error before sending all of "read responses" into the stream. In this case it is conceivable we could still have received a transaction ID even though the read operation finally rejects. With a bit of reworking we could capture the transaction ID and roll back the transaction even when the read rejects.

What are your thoughts on that?

Reply:
The `retryTransaction` is more of a hint to backend, that can help process the transaction more efficiently. It won't create adverse behaviour if previous transaction is still active.

Reply:
Ok I've reverted the conditional rollback completely and made rollback complete asynchronously.

Reply:
Ok I've reverted the conditional rollback completely and made rollback complete asynchronously.